syntax = "proto3";

package xroute.net_ordering;

message Node {
  // maze index
  sint32 maze_x = 1;
  sint32 maze_y = 2;
  sint32 maze_z = 3;

  // point coordinates
  sint32 point_x = 4;
  sint32 point_y = 5;
  sint32 point_z = 6;

  // node details
  NodeType type = 7;
  bool is_used = 8;
  uint32 net = 9; // the value starts from 1, 0 means not belongs to a net.
  uint32 pin = 10; // ditto.
}

enum NodeType {
  BLOCKAGE = 0;
  NORMAL = 1;
  ACCESS = 2;
}

message Initial {
  uint32 task_id = 1;
}

message Request {
  // dimensions of the maze
  uint32 dim_x = 1;
  uint32 dim_y = 2;
  uint32 dim_z = 3;

  repeated Node nodes = 4;

  uint32 reward_violation = 5;
  uint32 reward_wire_length = 6;
  uint32 reward_via = 7;
}

message Response {
  uint32 net_index = 1;
}

message Done {
  uint32 reward_violation = 5;
  uint32 reward_wire_length = 6;
  uint32 reward_via = 7;
}

message Message {
  oneof wrapper {
    Initial initial = 1;
    Request request = 2;
    Response response = 3;
    Done done = 4;
  }
}
